### DISCUSSION TITLE
Standardization of Rust Docs

### DISCUSSION THREAD

**Discussion by plebhash** [stratum-mining/stratum#970] [General]
Proposals on standardization of formatting the comments for Rust Docs

**Comment by rrybarczyk** [comment]
## Naming Conventions
Use clear and descriptive names for modules, functions, variables, and types. Follow Rust's naming conventions:
- **Modules**: snake_case
- **Types and Traits**: CamelCase
- **Functions and Variables**: snake_case
- **Constants and Statics**: SCREAMING_SNAKE_CASE

## Functions/methods:

```rs
/// This is a short description of the function.
/// 
/// Optional longer notes/dev notes. This block should be separated by one new line from the short description above.
/// 
/// # Arguments
/// * `arg_0`: This is a short description of the first argument of the function.
/// * `arg_1`: This is a short description of the first argument of the function.
/// 
/// # Returns
/// This is a description of the return type of the function on success and on error. This should include what could cause an error as well. It should be separated by one new line from the arguments section, or if no arguments, from the description. If no return type, do not include this section.
fn my_function(arg_0: u8, arg_1: &str) -> Result<()> {
  // Function implementation here
}
```

## Top level module
```rs
//! # Short Module Title
//!
//! This is a description of the purpose of the module. (Probably need more standards here, just kicking off the convo).
```

## Structs
Note that each member field is separated by a new line. Do we like this?

```rs
/// Short description of purpose.
///
/// Optional longer description/notes/dev notes. This should be separated by one new line from the short description of the struct.
struct MyStruct {
  /// Short description of field. If the field is very obvious, can exclude any comments.
  ///
  /// Optional longer description/notes/dev notes. This should be separated by one new line from the short description of the member field.
  field_0: u8,

  field_1: &str,  // Assume this is an obvious field, so no comments needed.

  /// Short description.
  field_3: CustomStruct,
}
```

**Reply by Fi3** [reply]
I find very useful when comment contains doc test. Also for function I would not require ARGS and RETURN section. Nice to have but most times are redundant. I would require the description section. If the function return a Result or can fail I would require to write in the comment how and why. (Ofc we can not be to pedantic here for example if a function allocate we shouldn't say in the comment that the allocation can fail)

**Reply by Fi3** [reply]
I'm not 100% against requiring args and return section, I think that will generate a lot of redundant docs. But better to have it then not have it when needed.

**Reply by Fi3** [reply]
I would require a doc test for every function where is possible to have one, that I'm 100% on

**Reply by jbesraa** [reply]
I am also not a big fan of the args/return in docs. When something unique or unusual in the args/ret or different error flows, it can be mentioned, otherwise, I agree with above comment that it feels redundant.

Regarding doc tests im on the fence. I would prefer to maintain unit tests plus integration tests to make sure all is working, and have a strong requirement for testing when reviewing PR. If we dont do that, I can see how doc tests can be useful.

**Reply by rrybarczyk** [reply]
Ok, then let's ditch the args/return. And I like the doc test idea a lot.

Something like this instead?

```
/// This is a short description of the function.
/// 
/// Optional longer notes/dev notes. This block should be separated by one new line from the short description above.
///
/// # Example (this is optional)
/// ```rust
/// use my_crate::my_function;
/// let result = my_function(0, "arg1");
/// assert!(result.is_ok());
/// ```
fn my_function(arg_0: u8, arg_1: &str) -> Result<()> {
  // Function implementation here
}
```

**Reply by rrybarczyk** [reply]
What about a doc test in the docs of the top of a module, overkill?

```
//! # Short Module Title
//!
//! This is a description of the purpose of the module.
//!
//! ## Example
//!
//! ```
//! use my_crate::add;
//! 
//! let result = add(2, 3);
//! assert_eq!(result, 5);
//! ```
```

**Reply by bit-aloo** [reply]
Doc test can only be added to public facing modules. Makes sense to have them.

**Reply by rrybarczyk** [reply]
Agree. Let's be sure to have an example on the top of every module (or at least where it makes sense). Then we will _not_ have examples anywhere else in the code (unless there is a specific need).

**Reply by plebhash** [reply]
while reviewing #1040 I noticed instances where the module headers were introducing a `## Types`

that is redundant since all types are already listed on Rust Docs automatically

doing this is actually a potential footgun, because now whenever we introduce code changes we need to:
- rewrite the docs immediately above the type
- rewrite the docs on the header (which can be easily overlooked)

**Comment by rrybarczyk** [comment]
Added the Naming Conventions in the main discussion comment.

**Comment by rrybarczyk** [comment]
Also, we can use attributes for documentation if there is docs that are getting too verbose. `#[doc(include = "path/to/file.md")]` allows for the ability to include an external file in the documentation.

**Comment by rrybarczyk** [comment]
I would like to also add that wherever possible, use the `cargo doc` supported type linking syntax to make the datatypes used in the crates clickable, routing you to that type's definition.

For example, consider writing Rust docs in the`noise_sv2::initiator` module, we can use the following syntax to achieve clickable reference links: 

```rs
/// By adding the brackets and back ticks around the [`Initiator`] struct, this word/type generated
/// in the cargo docs will be clickable, taking you directly to that type definition.
///
/// Since the `initiator` module imports `aes_gcm::KeyInit`, I can also simply reference this type
/// with [`KeyInit`], and it will take me to its definition in the `aes_gcm` crate docs.
///
/// If I need to reference a data type in the `initiator` docs that is not directly used in the
/// module, then I can reference it by giving the docs the full crate path as [`crate::Responder`]
/// or [`crate::aed_cipher::AeadCipher`]. These will not become a clickable reference link to these
/// types.
```

**Reply by rrybarczyk** [reply]
Please let me know if everyone agrees and I will move this to the main comment in this discussion to be a guideline.

**Reply by Fi3** [reply]
ack

**Reply by jbesraa** [reply]
Strong ack. Here is a nice example from rust-lightning https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/offers/nonce.rs#L21

**Comment by plebhash** [comment]
# module visibility

a [quote](https://github.com/stratum-mining/stratum/pull/1040/files#r1758464970) from @Fi3 :
> if a module is private, we don't have to document it for user but for maintainers.
>
> a rule of thumb is, if you think that you need to document a private module for the user, maybe the module should not be private

if the module (or `struct`, `enum`, `fn`, `type`, etc) is private and we still want to document it for maintainer eyes, we should use [Non Doc comments](https://doc.rust-lang.org/reference/comments.html#non-doc-comments), which are essentially  `//` inlines and `/* ... */` blocks

Non Doc comments are not rendered via `cargo doc` commands

**Comment by plebhash** [comment]
# examples

examples showing how to use public APIs are always good: they allow the reader to get a quick understanding of how the APIs should be used without having to read through the description of every single element

we might want to write an example showing how to use a `fn`, `struct`, `trait`, or maybe an entire `module` or `lib`

small examples (few lines of code) should be written as [doc tests](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html)

if the example is big, we should:
- create an `examples` directory inside the crate
- place each example as a `fn main() { ... }` inside a `examples/*.rs` file
- add this example to CI

---

examples should only cover public APIs

**Reply by rrybarczyk** [reply]
What do you all think of the syntax and verbage in [these comments](https://github.com/stratum-mining/stratum/blob/251e405b706392fa078b8f439ed0e5d54b009981/protocols/v2/codec-sv2/examples/encrypted.rs#L1C1-L17C7) in the `codec-sv2` example?

```rs
// # Using Sv2 Codec with Noise Encryption
//
// This example demonstrates how to use the `codec-sv2` crate to encode and decode Sv2 frames
// with Noise protocol encryption over a TCP connection. It showcases how to:
//
// * Perform a Noise handshake between the sender and receiver.
// * Create an arbitrary custom message type (`CustomMessage`) for encryption/encoding and
//   decryption/decoding.
// * Encode the message into an encrypted Sv2 frame using Noise.
// * Send the encrypted frame over a TCP connection.
// * Decode the encrypted Sv2 frame on the receiving side after completing the Noise handshake.
//
// ## Run
//
// ```
// cargo run --example encrypted --features noise_sv2
// ```
```

**Comment by plebhash** [comment]
# README.md

let's try to follow the format below:

```
# crate_name_sv2

[![crates.io](https://img.shields.io/crates/v/crate_name_sv2.svg)](https://crates.io/crates/crate_name_sv2)
[![docs.rs](https://docs.rs/crate_name_sv2/badge.svg)](https://docs.rs/crate_name_sv2)
[![rustc+](https://img.shields.io/badge/rustc-1.75.0%2B-lightgrey.svg)](https://blog.rust-lang.org/2023/12/28/Rust-1.75.0.html)
[![license](https://img.shields.io/badge/license-MIT%2FApache--2.0-blue.svg)](https://github.com/stratum-mining/stratum/blob/main/LICENSE.md)
[![codecov](https://codecov.io/gh/stratum-mining/stratum/branch/main/graph/badge.svg)](https://app.codecov.io/gh/stratum-mining/stratum/tree/main/protocols%2Fv2%2Fcrate_name_sv2)

crate description xxx
(Suggestion: Use similar language in the top of the crate's `libs.rs` module doc comments)

## Key Capabilities

description of key capabilities of the crate

this section should be optional, to be added only if there's meaningful information to be conveyed

## Main Components

list of main components of the crate

this section should be optional, to be added only if there's meaningful information to be conveyed

## Usage

basic instructions on how to add the crate to a project

> To include this crate in your project, run:
>
> ```bash
> cargo add <crate-name>
> ```

if the crate has feature flags, they should be described here

## Examples

description of examples, if any files under `examples` directory
```

**Reply by plebhash** [reply]
`const_sv2` reference by @GitGab19 https://github.com/GitGab19/stratum/tree/add-rust-docs-const-sv2/protocols/v2/const-sv2#readme

**Reply by plebhash** [reply]
one important detail to take into consideration:

`README.md` is what gets rendered on `crates.io`, so it's essentially the "face" of the crate

**Reply by rrybarczyk** [reply]
I am loving @GitGab19's `README.md`. 

Thoughts:
* One syntax thing I feel strongly about is enforcing a maximum line character length of 100 with exceptions being if links run over, or markdown tables, or similar.
* Do we want to link to the crates.io documentation somewhere?
For binary crates only, I think we should add run examples. For example, if we are writing a `README.md` for the translation proxy, we should explain and show each run command to start up a CPU miner and the pool role using the `pool` crate.
* Do we want to add anything in `README.md` regarding examples? Or should we just let the `examples/` directory speak for itself and perhaps add comments at the top of the file that includes the run command, ensuring that if specific feature flags are required in the run command, that it is clear to the user to use them. Perhaps we need a brief standard on top level example scripts?

**Reply by plebhash** [reply]
yeah if the crate has an `examples` directory (e.g.: `codec_sv2`) it makes sense to highlight that in the `README.md`

**Comment by plebhash** [comment]
# FFI

`protocols/v2/sv2-ff2/sv2.h` is committed to `sv2-ffi` crate, which is dynamically generated via `cbindgen`

`sv2-header-check.yaml` CI exists to make sure we always keep `protocols/v2/sv2-ff2/sv2.h` up-to-date

every time we update Rust Docs on some crate that is covered by `sv2-ffi`, some comments from this PR are expected to trickle down to the `sv2.h` that is dynamically generated via CI

therefore we should also update  `protocols/v2/sv2-ff2/sv2.h` when necessary, to avoid breaking CI

a new updated `sv2.h` can be generated via `scripts/build_headers.sh`, which should then be used to replace `protocols/v2/sv2-ff2/sv2.h`

---

one might object: "a Rust Docs PR should not touch any crates other than the one being documented"

I believe this is an exception, since these changes are atomically linked and therefore deserve to belong to the same PR

it should suffice to add a note to the PR description highlighting that `protocols/v2/sv2-ff2/sv2.h` is being updated with the new comments

**Comment by plebhash** [comment]
# crate versioning

the main point of writing the Rust Docs is to have them publicly displayed, so we need to publish those crates to `crates.io` once the Rust Docs are written

`cargo publish` will not work unless the crate version being published is bigger than the latest published version

for `framing_sv2`, `codec_sv2`, `const_sv2` and `noise_sv2`, we bumped MINOR

it was pointed out on https://github.com/stratum-mining/stratum/pull/1222 that we should probably start bumping PATCH instead

**Reply by plebhash** [reply]
like I said on #1222, I'm not opinionated on MINOR vs PATCH

we initially bumped MINOR (since there was no bug being fixed), but the argument that there's no new API features being added (and therefore PATCH should be bumped) also makes sense to me

whatever we decide to go for, we should align so we do the same thing for the crates we'll document on this next iteration

**Reply by GitGab19** [reply]
Even if it's not a real feature addition, it's still an improvement of the crate. So I would bump MINOR since the PATCH clearly means something has been fixed, and it's not the case

**Reply by Fi3** [reply]
no cargo use semantic version so if you bump minor it will expect that there are new feature and it will keep it in consideration when it have to chose which versions use. So bumping minor when there are no changes is just misleading cargo.

**Reply by Fi3** [reply]
patch means nothing in the API is changed, and that is the case

**Reply by GitGab19** [reply]
Do you have any reference to this? I'm ok with both tbh

**Reply by Fi3** [reply]
https://doc.rust-lang.org/cargo/reference/resolver.html
https://doc.rust-lang.org/cargo/reference/semver.html

**Reply by Fi3** [reply]
> When multiple packages specify a dependency for a common package, the resolver attempts to ensure that they use the same version of that common package, as long as they are within a SemVer compatibility range. It also attempts to use the greatest version currently available within that compatibility range. For example, if there are two packages in the resolve graph with the following requirements:

**Reply by GitGab19** [reply]
I don't think there would be any particular problem with bumping MINOR, but I'm totally fine with bumping the PATCH in this case 👍

**Reply by Fi3** [reply]
If you bump minor with no API changes you are not following semver this should be enough for not doing it.

But below a pratical example where this would be an issue:

let's say that you bump minor for A and you have 1.1.0 and 1.2.0
Now if you have crate B that depend upon =1.1.0 and crate C that depend on ~1.2.0 and crate D use both crate B and C you have a problem. You will have to import at least 2 different version when you could have imported just one. Then you could have  a bigger problem since types from A used by B are different then the ones used by C. This is just an example between uncountable cases like that.

**Reply by GitGab19** [reply]
Got it. So let's bump PATCH when we document crates 👍
